### PR TITLE
drivers: sdhc: Cadence SDHC driver bug fixes and other improvements

### PR DESCRIPTION
**Pull request by tanmaykathpalia**
* Internal Clock not getting stable during initialization
* SD clock was not enabled
* Removed duplicate file.
* Removed write to read-only bits
* Moved function pointer structure from the data section to the RO section
* Wrong macro names
* Refactoring of lower layer header file.

Signed-off-by: Tanmay Kathpalia <tanmay.kathpalia@intel.com>

### CODE REVIEW

**Comment by tanmaykathpalia** [comment]
@decsny @danieldegrasse Kindly review and provide your valuable feedback.

**Comment by tanmaykathpalia** [comment]
Some duplicate files were added in [64036](https://github.com/zephyrproject-rtos/zephyr/pull/64036) which had been taken care of in this PR.

**Comment by danieldegrasse** [comment]
@roymurlidhar can you take a look at this PR as well? Changes look ok to me but would appreciate a look from someone familiar with the IP itself

File: drivers/sdhc/sdhc_cdns_ll.h [hunk 1]
```diff
@@ -55,26 +65,38 @@
...
+#define CDNS_SRS12_TC			BIT(1)
+#define CDNS_SRS12_EINT			BIT(15)
+
+/* SDMA Buffer Boundary */
+#define BUFFER_BOUNDARY_4K		0U
+#define BUFFER_BOUNDARY_8K		1U
+#define BUFFER_BOUNDARY_16K		2U
+#define BUFFER_BOUNDARY_32K		3U
+#define BUFFER_BOUNDARY_64K		4U
+#define BUFFER_BOUNDARY_128K	5U
```
Comment (danieldegrasse) [inline drivers/sdhc/sdhc_cdns_ll.h:84]: Nit- align these indentations please

File: drivers/sdhc/sdhc_cdns_ll.h [hunk 2]
```diff
@@ -22,31 +23,40 @@
...
 
-/* SRS10 */
-#define CDNS_SRS10_DTW			1
-#define CDNS_SRS10_EDTW			5
-#define CDNS_SRS10_BP			8
-#define CDNS_SRS10_BVS			9
+/* SRS10 - Host Control 1 (General / Power / Block-Gap / Wake-Up) */
+#define LEDC				BIT(0)
+#define DT_WIDTH			BIT(1)
+#define HS_EN				BIT(2)
```
Comment (danieldegrasse) [inline drivers/sdhc/sdhc_cdns_ll.h:33]: Nit- align these indentations

File: drivers/sdhc/sdhc_cdns_ll.h [hunk 3]
```diff
@@ -170,64 +192,65 @@
...
 
 /* HRS16 */
-#define CDNS_HRS16_WRCMD0_DLY(x)	(x)
-#define CDNS_HRS16_WRCMD1_DLY(x)	((x) << 4)
-#define CDNS_HRS16_WRDATA0_DLY(x)	((x) << 8)
-#define CDNS_HRS16_WRDATA1_DLY(x)	((x) << 12)
+#define CDNS_HRS16_WRCMD0_DLY(x)		(x)
+#define CDNS_HRS16_WRCMD1_DLY(x)		((x) << 4)
+#define CDNS_HRS16_WRDATA0_DLY(x)		((x) << 8)
+#define CDNS_HRS16_WRDATA1_DLY(x)		((x) << 12)
```
Comment (danieldegrasse) [inline drivers/sdhc/sdhc_cdns_ll.h:245]: Unclear why this indentation was added?

File: drivers/sdhc/sdhc_cdns_ll.h (hunk 1 above)
Comment (tanmaykathpalia) [reply drivers/sdhc/sdhc_cdns_ll.h]: corrected.

**Review by decsny — DISMISSED** [review]

**Comment by danieldegrasse** [comment]
@tanmaykathpalia do you know if @roymurlidhar will be able to review these changes? I'd appreciate confirmation from them that everything looks good here, since I am not familiar with the Cadence IP

**Comment by tanmaykathpalia** [comment]
> @tanmaykathpalia do you know if @roymurlidhar will be able to review these changes? I'd appreciate confirmation from them that everything looks good here, since I am not familiar with the Cadence IP
@danieldegrasse I spoke with @roymurlidhar, and he mentioned that he would try to review it tomorrow. Also, he left Intel, so he won't have access to IP documents anymore. Let's wait for tomorrow.

**Comment by tanmaykathpalia** [comment]
@decsny thanks for your time and approval.

File: drivers/sdhc/sdhc_cdns_ll.c [hunk 4]
```diff
@@ -720,8 +721,9 @@ static int sdhc_cdns_send_cmd(struct sdmmc_cmd *cmd, struct sdhc_data *data)
 
 	timeout = CARD_REG_TIME_DELAY_US;
 	if (!WAIT_FOR(((((sys_read32(cdns_params.reg_base + SDHC_CDNS_SRS12)) &
-		CDNS_SRS03_CMD_DONE) == 1) | (((sys_read32(cdns_params.reg_base +
-		SDHC_CDNS_SRS12)) & ERROR_INT) == ERROR_INT)), timeout, k_busy_wait(1))) {
+			CDNS_SRS12_CC) == CDNS_SRS12_CC) | (((sys_read32(cdns_params.reg_base +
+			SDHC_CDNS_SRS12)) & CDNS_SRS12_EINT) == CDNS_SRS12_EINT)),
+			timeout, k_busy_wait(1))) {
```
Comment (pbalsundar) [inline drivers/sdhc/sdhc_cdns_ll.c:726]: can we use k_msleep instead of k_busy_wait?

Comment (tanmaykathpalia) [reply drivers/sdhc/sdhc_cdns_ll.c]: done

**Review by pbalsundar — APPROVED** [review]

**Comment by ROYMURLI** [comment]
Hi, I have reviewed the changes and it looks fine for me. We can go ahead and approve this PR. Thanks :)

**Review by danieldegrasse — APPROVED** [review]

**Comment by decsny** [comment]
> @decsny can you approve again and merge the PR.

i dont have merge rights

**Review by decsny — APPROVED** [review]

**Pull request merged**